src/peft/tuners/oft/layer.py

```diff
@@ -14,106 +14,293 @@
 
 import math
 import warnings
-from typing import Any, List, Optional, Set, Tuple
+from typing import Any, List, Optional, Set, Union
 
 import torch
 import torch.nn as nn
+import torch.nn.functional as F
 
-from peft.tuners.lycoris_utils import LycorisLayer, check_adapters_to_merge
+from peft.tuners.tuners_utils import BaseTunerLayer, check_adapters_to_merge
 
 
-class OFTLayer(nn.Module, LycorisLayer):
+class MultiplicativeDropoutLayer(nn.Module):
+    """
+    Implements the multiplicative dropout layer for BOFT.
+    """
+
+    def __init__(self, p=0.0):
+        """
+        Initializes the multiplicative dropout layer.
+
+        Parameters:
+        p (float): The probability of dropping out a block. Defaults to 0.0.
+        """
+        super().__init__()
+        self.p = p
+
+    def forward(self, x):
+        """
+        Applies multiplicative dropout to the input tensor.
+
+        Parameters:
+        x (Tensor): The input tensor of shape (D, H, H), where `D` represents
+                    the number of OFT blocks, and `H` is the size of the square
+                    blocks along the last two dimensions, the block size in OFT.
+        """
+        if self.training:
+            # Ensure the last two dimensions are the same
+            if x.shape[-1] != x.shape[-2]:
+                raise ValueError("The last two dimensions of input should be the same!")
+
+            D, H, _ = x.shape
+
+            # If block share, skip the multiplicative dropout
+            if D == 1:
+                return x
+
+            # Create a mask with 1s for matrices to be replaced with identity and 0s otherwise
+            num_to_replace = int(self.p * D)
+            num_zeros = D - num_to_replace
+
+            # Generate a flat tensor with desired number of 1s and 0s
+            mask = torch.cat([torch.ones(num_to_replace, device=x.device), torch.zeros(num_zeros, device=x.device)])
+
+            # Shuffle and reshape the mask
+            mask = mask[torch.randperm(D)].view(D, 1, 1)
+
+            # Use the mask to combine original matrices and identity matrices
+            eye_matrix = torch.eye(H, device=x.device).repeat(D, 1, 1)
+            x = (1 - mask) * x + mask * eye_matrix
```

Comment:
Just for my understanding, do we need some kind of scaling like 1/(1 - p) in normal dropout to ensure that there is no scale shift between training and inference?

Reply:
Thanks for the comment, actually this catched by self.training, meaning in inference there is never multiplicative dropout applied. (It will set certain blocks as identity and cannot be trained during inference.) Best.